File: johnson.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace TTCS_THUAT_TOAN_TIM_DUONG_DI_NGAN_NHAT
{
    public class Diem_G /// them diem G trong so cua G voi diem dau = 0
    {
        public NhapDothi
         {FileStream f = new FileStream("D:\\HocTap\\dulieu1.inp", FileMode.Open);
            StreamReader rd = new StreamReader(f, Encoding.UTF8);
            string dls = rd.ReadToEnd();
            Console.WriteLine(dls);
            rd.Close();
            int[,] duLieu = new int[50, 50];
            this.duLieu = duLieu;
            khoitao();
        }
        void khoitao()
        {
            listDuongDi = new List<DuongDi>();
            daXet = new List<bool>();
            for (int i = 0; i < duLieu.GetLength(0); i++) daXet.Add(false);

        }
    
    }
    public class Bellman /// dung bell man de tinh trong so
    {
          class Canh
    {
        public string u;
        public string v;
        public int cost;
        public Canh(string _u,string _v,string _cost)
        {
            u = _u;
            v = _v;
            cost = int.Parse(_cost);
        }
    }

    class DoThi
    {
        List<string> dsDinh;
        List<Canh> dsCanh;
        string nguon;
        
        public DoThi(string path)
        {
            string[] data = System.IO.File.ReadAllLines(path);
            dsCanh = new List<Canh>();
            dsDinh = new List<string>();
            for(int i = 0; i < data.Length; i++)
            {
                string[] s = data[i].Split(' ');
                dsCanh.Add(new Canh(s[0], s[1], s[2]));
                if(!dsDinh.Contains(s[0]))
                    dsDinh.Add(s[0]);
                if (!dsDinh.Contains(s[1]))
                    dsDinh.Add(s[1]);
            }
            Console.WriteLine("Danh sach cac dinh: ");
            for(int i = 0; i < dsDinh.Count; i++)
            {
                Console.Write(dsDinh[i] + " ");
            }
            Console.Write("Chon dinh bat dau : ");
            nguon = Console.ReadLine();
        }
        public String Print()
        {
            return "Dinh nguon la: " + nguon;
        }

        public void BellmanFord()
        {
            //distance[], predecessor[]
            Dictionary<string, int> distance = new Dictionary<string,int>();
            Dictionary<string, string> predecessor = new Dictionary<string,string>();
            /*
             Bước 1: Khởi tạo đồ thị
                    Cho mỗi đỉnh trong danh sách các đỉnh
                    distance[v] = inf           // Tại bước đầu tiên, các đỉnh có độ dài là vô cùng
                    predecessor[v] = inf        // Và không có bất kì đỉnh tiền nhiệm nào
             */
            //Xác định inf (vô cùng)
            int inf = dsCanh.Max(p => p.cost) + 100;
            foreach(string v in dsDinh)
            {
                distance.Add(v, inf);
                predecessor.Add(v, null);
            }
            // distance[source] = 0                   // Diem xuat luon co do lon bang 0
            distance[nguon] = 0;
            /* bước 2: kết nạp cạnh
                for i from 1 to size(danh_sách_đỉnh):       
                for each(u, v) in danh_sách_cung:
                if khoảng_cách(v) > khoảng_cách(u) + trọng_số(u, v):
                khoảng_cách(v):= khoảng_cách(u) + trọng_số(u, v)
                đỉnh_liền_trước(v):= u*/
            for(int i = 0; i < dsDinh.Count -1; i++)
            {
                foreach(Canh c in dsCanh)
                {
                    if(distance[c.u]+c.cost < distance[c.v])
                    {
                        distance[c.v] = distance[c.u] + c.cost;
                        predecessor[c.v] = c.u;
                    }
                }
            }
            // bước 3: kiểm tra chu trình âm
            /*for each(u, v) in danh_sách_cung:
            if khoảng_cách(v) > khoảng_cách(u) + trọng_số(u, v):
            error "Đồ thị chứa chu trình âm"*/
            foreach(Canh c in dsCanh)
            {
                if(distance[c.u] + c.cost < distance[c.v])
                {
                    Console.WriteLine("Do thi co chua chu trinh am");
                }
            }
            // Trả vể độ lớn và các đỉnh nguồn
            //in kết quả
            Console.WriteLine("Duong di ngan nhat tu dinh" + nguon + "la: ");
            Console.WriteLine("--Khoang cach tu nguon den cac dinh: ");
            foreach(var item in distance)
            {
                Console.WriteLine(item.Key + " " + item.Value);
            }
            Console.WriteLine("--Duong di giua cac dinh: ");
            Console.WriteLine("Dinh\tTien nhiem");
            foreach(var item in predecessor)
            {
                Console.WriteLine(item.Key + "  " + item.Value);       
            }

        }
    }


    }
    public class WeightNew// luu lai do thi moi 
    {
        int johnson()
        {
	int s = n;				
	int u = 0, v = 0;
	for(; u < n; u++){
		add_arc(s,u,0);			
	}
	int *P = bellman_ford(G,s,n+1);		
	for(u = 0; u < n; u++){
		vlist *nbrs = G[u];		
		while(nbrs != NULL){
			v = nbrs->v;
			nbrs->Delta = nbrs->w + P[u] - P[v];	
			nbrs = nbrs->next;		
		}
	}
	Console.WriteLine(" Trong so thay doi \n");
//////////////// in list tai day
	H = (int *)malloc((n-1)*(sizeof(int)));		
	pos = (int *)malloc((n-1)*sizeof(int));
	
	int **D = (int **) malloc(n*sizeof(int*));		
	d = (int *)malloc(n*sizeof(int));			
	for( u = 0; u < n; u++){
		D[u] = (int *)malloc(n*sizeof(int));		
		dijkstra_heap(G, u, n);	 	
		for(v = 0; v < n; v++){
			D[u][v] = d[v] + P[v] - P[u];	
		}
	}
	
	return D;
}
    }
    public class TimDuongDijktra /// su dung code cua Long gia quyet Dijktra
    {
        public int[,] duLieu { get; private set; }
        List<bool> daXet { get; set; }
        List<DuongDi> listDuongDi { get; set; }
        int diemDauM { get; set; }
        int diemCuoiM { get; set; }
        public TimDuongDijktra()
        {
           // du lieu moi nhan tu WeightNew
            khoitao2();
        }
        public TimDuongDijktra(int[,] duLieu)
        {
            if (duLieu != null)
            {
                this.duLieu = duLieu;
                khoitao2);
            }
        }
    
        void khoitao2()/// nhan do  thi sau khi thay doi trong so
        {
            listDuongDi = new List<DuongDiMoi>();
            daXet = new List<bool>();
            for (int i = 0; i < duLieu.GetLength(0); i++) daXet.Add(false);

        }
        public List<int> TimDuong(int diemDauM, int diemCuoiM)
        {
            khoitao2();

            diemDauM--;
            diemCuoiM--;
            this.diemDauM = diemDauM;
            this.diemCuoiM = diemCuoiM;
            if (diemDauM < 0 || diemDauM > duLieu.GetLength(0) || diemCuoiM < 0 || diemCuoiM > duLieu.GetLength(0)) return null;

            List<int> duong = new List<int>();


            DuongDi dd = new DuongDi();
            dd.diemDaDi.Add(diemDauM);

            this.listDuongDi.Add(dd);

            int dangXet = diemDauM;
            daXet[diemDauM] = true;

            while (!KetThuc(diemCuoiM))
            {
                DuongDi timDuong = this.listDuongDi[0];
                int diemMoi = -1;

                foreach (DuongDi item in this.listDuongDi)
                {
                    int s = TimDuong(item);
                    if (s != -1)
                    {
                        if (diemMoi == -1)
                        {
                            diemMoi = s;
                            timDuong = item;
                        }
                        else
                        {
                            if (s < diemMoi)
                            {
                                diemMoi = s;
                                timDuong = item;
                            }
                        }
                    }
                }

                int soLuongThem = TimDiemKe(timDuong.diemDaDi[timDuong.diemDaDi.Count - 1]).Count;
                for (int i = 0; i < soLuongThem - 1; i++)
                {
                    DuongDi ddn = new DuongDi();
                    foreach (int item in timDuong.diemDaDi) ddn.diemDaDi.Add(item);

                    this.listDuongDi.Add(ddn);
                }

                if (diemMoi != -1)
                {
                    timDuong.diemDaDi.Add(diemMoi);
                    daXet[diemMoi] = true;
                }


                if (diemMoi == diemCuoiM) duong = timDuong.diemDaDi;
            }

            return duong;
        }
        bool KetThuc(int diemCuoi)
        {
            bool kt = true;

            foreach (DuongDi item in listDuongDi)
            {
                if (!item.isDie)
                {
                    kt = false;
                }
            }

            foreach (DuongDi item in listDuongDi)
            {
                if (!item.isDie)
                {
                    if (item.diemDaDi[item.diemDaDi.Count - 1] == diemCuoi) return true;
                }
            }

            return kt;
        }

        int TimDuong(DuongDi dd)
        {
            int le = dd.diemDaDi.Count;
            int diem = dd.diemDaDi[le - 1];
            List<int> list = TimDiemKe(diem);

            if (list.Count == 0)
            {
                dd.isDie = true;
                return -1;
            }

            int min = duLieu[diem, list[0]];
            int diemToi = list[0];
            foreach (int item in list) if (min >= duLieu[diem, item] && duLieu[diem, item] > 0)
                {
                    if (diem == this.diemCuoiM && min > duLieu[diem, item]) continue;
                    min = duLieu[diem, item];
                    diemToi = item;
                }

            return diemToi;
        }

        List<int> TimDiemKe(int diem)
        {
            List<int> listkq = new List<int>();

            for (int i = 0; i < duLieu.GetLength(0); i++)
            {
                if (diem != i && duLieu[diem, i] > 0 && !daXet[i]) listkq.Add(i);
            }

            return listkq;
        }
        public int TinhDuong(List<int> duongDi)
        {
            if (duLieu != null)
            {
                if (duongDi.Count == 0)
                {
                    return 0;
                }
                else
                {
                    int quangDuong = 0;

                    for (int i = 0; i < duongDi.Count - 1; i++)
                    {
                        quangDuong += duLieu[duongDi[i], duongDi[i + 1]];
                    }

                    return quangDuong;
                }
            }
            return 0;
        }
    }

    public class DuongDi
    {
        public List<int> diemDaDi { get; private set; }

        public bool isDie { get; set; }
        public DuongDi()
        {
            diemDaDi = new List<int>();
            this.isDie = false;

        }
        public bool ThemDiem(int diem)
        {
            foreach (int item in this.diemDaDi) if (item == diem) return false;
            this.diemDaDi.Add(diem);
            return true;
        }
    }
    class Program
    {
        static void Main(string[] args)
        {
              WriteLine.Console("Nhap diem bat dau:");// Nhap diem G vs trong so 0 vao dinh bat ban muon
        readline.Console("%d",&r);
        graph.edge[r].src = 0; 
        graph.edge[r].dest = r; 
        graph.edge[r].weight = 0; 
  
            TimDuongDijktra tt = new TimDuongDijktra();
            tt.TimDuong(1, 6);
        }
    }
}

File: johnson.cpp


#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<time.h>
#define INFTY 1000000000


int *H; 	
int hsize = 0; 	
int *pos;	 // mang danh dau vi tri cua moi dinh trog mang Heap
void up_heapify(int u);
int parent(int u);
void print_heap();
void decrease_key(int u, int K);
void down_heapify(int u);
int extract_min();
void build_heap(int n, int s);

typedef struct vlist{
	int v;		
	int w;		
	int Delta;	
	struct vlist *next;
}vlist;

vlist **G;			
int n;					
int **D;				
int *d;					
void read(); 				
void them_cung(int u, int v, int w);
int *bellman_ford(vlist **G, int s, int n);
void dijkstra_heap(vlist** G, int s, int n);
int  **johnson();

void inmang(int *A, int n);
void indothi();
void indscanh(int u);


int **johnson(){
	int s = n;				
	int u = 0, v = 0;
	for(; u < n; u++){
		them_cung(s,u,0);			
	}
	int *P = bellman_ford(G,s,n+1);		
	for(u = 0; u < n; u++){
		vlist *nbrs = G[u];		
		while(nbrs != NULL){
			v = nbrs->v;
			nbrs->Delta = nbrs->w + P[u] - P[v];	
			nbrs = nbrs->next;		
		}
	}
	printf("\nDo thi sau khi doi trong so\n");
	indothi();
	H = (int *)malloc((n-1)*(sizeof(int)));		
	pos = (int *)malloc((n-1)*sizeof(int));
	
	int **D = (int **) malloc(n*sizeof(int*));		
	d = (int *)malloc(n*sizeof(int));			
	for( u = 0; u < n; u++){
		D[u] = (int *)malloc(n*sizeof(int));		
		dijkstra_heap(G, u, n);	 	
		for(v = 0; v < n; v++){
			D[u][v] = d[v] + P[v] - P[u];	
		}
	}
	
	return D;
}

int *bellman_ford(vlist **G, int s, int n){
	int* D = (int *)malloc(n*sizeof(int));
	int i = 0;
	for(; i < n; i++)D[i] = INFTY;
	D[s] = 0;
	int v = 0;
	vlist *nbrs = G[s];	
	while(nbrs != NULL){	
		v = nbrs->v;
		D[v] = nbrs->w;
		nbrs = nbrs->next;
	}
	int  u = 0;
	for(i = 0; i < n-2; i++){
			for(u = 0; u < n; u++){
				vlist *nbrs = G[u];	
				while(nbrs != NULL){
					v = nbrs->v;
					if(D[v] > D[u] + nbrs->w){
						D[v] = D[u] + nbrs->w;
					}
				nbrs = nbrs->next;	
				} 
			}
	}
	return D;
}

void dijkstra_heap(vlist** G, int s, int n){
	int i = 0;
	for(; i < n; i++)d[i] = INFTY;
	d[s] = 0;
	int v = 0;
	vlist *nbrs = G[s];
	while(nbrs != NULL){	
		v = nbrs->v;
		d[v] = nbrs->Delta;
		nbrs = nbrs->next;
	}
	build_heap(n,s);
	int u;
	while(hsize != 0){ 	
		u = extract_min();
		vlist *nbrs = G[u];	
		while(nbrs != NULL){	
			v = nbrs->v;
			if(d[v] > d[u] + nbrs->Delta){
				d[v] = d[u] + nbrs->Delta;
				decrease_key(v,d[v]);
			}
			nbrs = nbrs->next;	
		} 
	}
}

// them dinh, canh, khoang cach
void read(){
	n = 6;
	int i = 0, j = 0;
	G = (vlist **) malloc((n+1)*sizeof(vlist*));	
	for(; i < n+1; i++)G[i] = NULL;
	them_cung(0,2,-1);	
	them_cung(1,0,5);		
	them_cung(2,0,3);		
	them_cung(2,1,-3);	
	them_cung(2,4,11);	
	them_cung(3,0,-6);	
	them_cung(4,3,-3);	
	them_cung(4,5,4);		
	them_cung(5,1,-4);	
	them_cung(5,2,1);		

}
// them G
void them_cung(int u, int v, int w){
	vlist* vnode = (vlist *)malloc(sizeof(vlist));
	vnode->v = v;
	vnode->w = w;
	vnode->Delta = -1;	
	vnode->next = G[u];
	G[u] = vnode;
}


void up_heapify(int u)// u không phai là goc cua heap
{
	int v = parent(u);
	if(v != -1 && d[H[u]] < d[H[v]]){
		int tmp = H[u];
		H[u] = H[v];
		pos[H[v]] = u;		
		H[v] = tmp;
		pos[tmp] = v;		
		up_heapify(v);
	}
}


//Hai nut con cua nut u la 2u + 1 và 2u + 2, nut nho hon là nut con ben trai
int parent(int u){
	return ((u&1)==0 ? ((u-2)>> 1) : (u-1) >> 1);
}



void in_dong(){
	int i = 0;
	for(; i < hsize; i++){
		printf("%d ", H[i]);
	}
	printf("\n");
}


void decrease_key(int u, int k){
	d[u] = k;
	up_heapify(pos[u]);
}

int extract_min(){
	int tmp  = H[0];
	H[0] = H[hsize-1];
	pos[H[0]] = 0;
	hsize--;
	down_heapify(0);
	return tmp;
}

void down_heapify(int u)// u không phai là lá cua dong
{
	int m = 2*u+1;
	if(m < hsize){	
		if(2*u+2 < hsize && d[H[m]] > d[H[2*u+2]]){
			m = 2*u+2;		
		}
		if(d[H[u]] > d[H[m]]){
			int tmp = H[u];
			H[u] = H[m];
			pos[H[m]] = u;	
			H[m] = tmp;
			pos[tmp] = m;	
			down_heapify(m);		
		}
	}
}

// Xây dung dong tu mot mang trong thoi gian O (n)
void build_heap(int n, int s){
	hsize = n-1;
	memset(H,-1,hsize*sizeof(int));		
	memset(pos, -1, hsize*sizeof(int));	
	int i = 0;
	for(; i < s; i++) {
		H[i] = i;
		pos[i] = i;
	}

	for(i = s+1; i < n; i++){
		H[i-1]=i;
		pos[i] = i-1;
	}
	for(i = hsize-1; i >=0; i--){
		down_heapify(i);
	}
}


void indothi(){
	int u = 0;
	for(; u < n; u++){
	indscanh(u);
	}	
}
void indscanh(int u)
{
	vlist *it = G[u];
	while(it != NULL){
		printf("(%d->%d,%d,%d) ",u,it->v, it->w, it->Delta);
		it = it->next;
	}
	printf("\n");
}
void inmang(int *A, int n){
	int i = 0;
	for(; i < n; i++){
		printf("%d ", A[i]);
	}
	printf("\n");

}
int main(){
	read();
	indothi();
	D = johnson();
	printf("\nTat ca khoang cach cua Johnson\n");
	int i = 0;
	for(; i < n; i++) inmang(D[i], n);
}
